consensus/src/core.rs

```diff
@@ -440,24 +469,143 @@ impl Core {
         Ok(())
     }
 
-    pub async fn run(&mut self) {
-        // Upon booting, generate the very first block (if we are the leader).
-        // Also, schedule a timer in case we don't hear from the leader.
+    #[cfg_attr(not(feature = "fast-sync"), allow(unused_variables))]
+    pub async fn run(&mut self, start_mode: CoreStartMode) {
+        // Log the received mode IMMEDIATELY for debugging
+        info!("Core::run entered with start_mode: {:?}", start_mode);
+        // --- Fast Sync Initialization Phase ---
+        // This block only runs if the 'fast-sync' feature is enabled AND
+        // the start_mode requests it.
+        #[cfg(feature = "fast-sync")]
+        if matches!(start_mode, CoreStartMode::FastSync) {
+            info!("Fast Sync: Core entering sync mode...");
+            const FAST_SYNC_TIMEOUT_MS: u64 = 10_000; // 10 second timeout
+
+            // 1. Request Checkpoint from a peer
+            // Simplification: Ask the first peer we know (excluding ourselves)
+            let peers = self.committee.broadcast_addresses(&self.name);
+            let mut checkpoint_fetched = false;
+            if let Some((_peer_name, peer_addr)) = peers.first() {
+                info!("Fast Sync: Requesting checkpoint from {}", peer_addr);
+                let request = ConsensusMessage::SyncCheckpointRequest(self.name);
+                match bincode::serialize(&request) {
+                    Ok(serialized_request) => {
+                        self.network.send(*peer_addr, Bytes::from(serialized_request)).await;
+                    }
+                    Err(e) => {
+                        error!("Fast Sync: Failed to serialize checkpoint request: {}", e);
+                        // Proceed to normal startup without checkpoint
+                    }
+                }
+
+                // 2. Wait for Response or Timeout
+                let deadline = sleep(Duration::from_millis(FAST_SYNC_TIMEOUT_MS));
+                tokio::pin!(deadline);
+
+                loop {
+                    tokio::select! {
+                        biased;
+                        // Check for incoming messages
+                        maybe_message = self.rx_message.recv() => {
+                            match maybe_message {
+                                Some(ConsensusMessage::SyncCheckpointResponse(Some(checkpoint_block))) => {
+                                    info!("Fast Sync: Received checkpoint block {}", checkpoint_block);
+
+                                    // --- PoC Verification ---
+                                    // WARNING: Very happy-path approach.
```

Comment:
Can you please explain what do you mean by this?

Reply:
Sure, so essentially I didn't come up with any solid ruling for checkpoint blocks, or committees, and just verified them against the (static) committee before loading the current epoch, round, best quorum and latest commit/vote rounds into core. From my understanding this is a very critical part and I think there's a lot of edge cases - hence the "PoC Verification" warning. It works for the "verification" level I need in this PoC (and for the fab local-fast-sync) which was my scope.